evalml/automl/engine/engine_base.py

```diff
@@ -304,26 +328,105 @@ def train_and_score_pipeline(
+                schema=False,
+                get_hashes=True,
+            )
+            stored_pipeline = full_pipeline
+            logger.debug(f"\t\t\tFull training data pipeline: finished training")
+            if (
+                automl_config.optimize_thresholds
+                and is_binary(automl_config.problem_type)
+                and full_pipeline.threshold is not None
+            ):
+                logger.debug(
+                    f"\t\t\tFull data pipeline: Optimal threshold found ({full_pipeline.threshold:.3f})"
+                )
+            logger.debug(f"\t\t\tScoring trained full training data pipeline")
+            holdout_scores = full_pipeline.score(
+                X_holdout,
+                y_holdout,
+                objectives=objectives_to_score,
+                X_train=full_X_train,
+                y_train=full_y_train,
+            )
+            logger.debug(
+                f"\t\t\tFull data pipeline: {automl_config.objective.name} score: {holdout_scores[automl_config.objective.name]:.3f}"
+            )
+            holdout_score = holdout_scores[automl_config.objective.name]
+            pipeline_cache[hashes] = full_pipeline.component_graph.component_instances
+        except Exception as e:
+            if automl_config.error_callback is not None:
+                automl_config.error_callback(
+                    exception=e,
+                    traceback=traceback.format_tb(sys.exc_info()[2]),
+                    automl=automl_config,
+                    fold_num=i,
+                    pipeline=pipeline,
+                )
+            if isinstance(e, PipelineScoreError):
+                nan_scores = {objective: np.nan for objective in e.exceptions}
+                holdout_scores = {**nan_scores, **e.scored_successfully}
+                holdout_scores = OrderedDict(
+                    {
+                        o.name: holdout_scores[o.name]
+                        for o in [automl_config.objective]
+                        + automl_config.additional_objectives
+                    }
+                )
+                holdout_score = holdout_scores[automl_config.objective.name]
+            else:
+                holdout_score = np.nan
+                holdout_scores = OrderedDict(
+                    zip(
+                        [n.name for n in automl_config.additional_objectives],
+                        [np.nan] * len(automl_config.additional_objectives),
+                    )
+                )
+        logger.info(
+            f"\tFinished holdout set scoring - {automl_config.objective.name}: {holdout_score:.3f}"
+        )
+
+    training_time = time.time() - start
     return {
         "scores": {
             "cv_data": cv_data,
             "training_time": training_time,
             "cv_scores": cv_scores,
             "cv_score_mean": cv_score_mean,
+            "holdout_score": None
+            if X_holdout is None and y_holdout is None
```

Comment:
We end up checking `if X_holdout is (not) None and y_holdout is (not) None` a lot in this function, and this part is a little hard to read - could you clean it up a little by saving a local variable, `use_holdout` or something along those lines?

Reply:
Moved check to `use_holdout` var